pkg/lib/fsinfo.ts

```diff
@@ -87,7 +87,8 @@ export class FsInfoClient extends EventEmitter<FsInfoEvents> {
     constructor(path: string, attrs: (keyof FileInfo)[], options?: cockpit.JsonObject) {
         super();
 
-        this.channel = cockpit.channel({
+        // HACK cockpit.js requires fsinfo for cockpit.file.watch() so we can't import cockpit as normal but only via `window`.
```

Comment:
Ah, that plus `import type` is the solution for circular import? Seems fine to me for `cockpit.channel`, which should be a stable API. I think it'd break bundling pkg/lib/cockpit.js though, so we need to be careful.

Reply:
`import type` can be a circular import, typescript does magic to make that work. The `import type` is stripped out once the code is converted to real JavaScript.

About breaking `pkg/lib/cockpit.js`, that is a good question.

Reply:
`window.cockpit` gets set in the main script body of `cockpit.js`, which should be imported by just about anything, and it should happen first.  We might imagine a theoretical user if only the fsinfo API which would break, if `cockpit.js` wasn't also included in its bundle.

In fact, we *almost* have one of those users right now, which is the fsinfo QUnit case.  The only reason that it wasn't broken by this change is because it first calls `cockpit.spawn()` to set up the directory for testing... if we didn't have that, then I'm quite sure that this hack would have broken.

That's a fairly unusual situation, though...  I suppose someone could create a react app that directly imported fsinfo and used it and only it.  Then we'd be in trouble.  But nobody is doing that... :shrug: 

We could wait until after #20666 to "do this right" if we wanted to.